### PR TITLE
[WIP] 4373 - document security events

### PR DESCRIPTION
**Pull request by kevintweber**
This is documentation for #4373.  This pull request has two parts:
- The component/security/authentication additions
- The new cookbook security cookbook page.

The new cookbook page is designed to give an example of how to use authentication events in an application.  I have given example code for two of the four classes needed to throttle failed login attempts.  The other two classes are a doctrine entity, and a class containing throttle domain logic, both of which are outside the scope of this documentation.

I would welcome any comments and suggestions.  Thank you.

### CODE REVIEW

**Comment by kevintweber** [comment]
Fudge.  I just realized that I used the RequestStack in the cookbook page, which was introduced in 2.4 not 2.3.  Alas...

File: components/security/authentication.rst [hunk 1]
```diff
@@ -267,5 +267,55 @@ in) is correct, you can use::
...
+security.interactive_login       ``SecurityEvents::INTERACTIVE_LOGIN``             :class:`Symfony\Component\Security\Http\Event\InteractiveLoginEvent`
+security.switch_user             ``SecurityEvents::SWITCH_USER``                   :class:`Symfony\Component\Security\Http\Event\SwitchUserEvent`
+===============================  ================================================  =========================================================================
+
+Authentication Success and Failure Events
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When a provider authenticates the user, a ``security.authentication.success``
+event is dispatched.  Likewise, when no providers authenticate the user,
+a ``security.authentication.failure`` event is dispatched.  You
```
Comment (weaverryan) [inline components/security/authentication.rst]: This sentence kind of makes it sound like this event will happen when no providers authentication versus "1 authenticator fails authentication". How about:

```rst
Likewise, when a provider attempts authentication but fails (i.e. throws an
``AuthenticationException``), a ``security.authentication.failure`` event is dispatched.
```

File: components/security/authentication.rst [hunk 2]
```diff
@@ -267,5 +267,55 @@ in) is correct, you can use::
...
+could listen on the ``security.authentication.failure`` event, for example,
+in order to log failed login attempts.
+
+.. tip::
+
+    One of the authentication events is always triggered when a request points
+    to a secured area.
```
Comment (weaverryan) [inline components/security/authentication.rst]: Is this true? I see these events thrown only in `AuthenticationProviderManager`. Is this called if no authentication listener is activated on a request - like if I'm already logged in, but am not at `/login_check` (assuming form login is my only method)? Also, if `anonymous` is activated, does that cause `security.authentication.success` to be dispatched on every request?

This stuff is tricky - all the more reason why I'd like to merge this PR.

File: cookbook/security/throttle_failed_login.rst [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        public function onAuthenticationFailureEvent(AuthenticationFailureEvent $event)
+        {
+            $this->repository->logAuthenicationFailure(
+                $event->getToken(),
+                $this->requestStack->getCurrentRequest()
+            );
+        }
+    }
+
+Now you are logging all the relevant details about every failed login attempt.
```
Comment (weaverryan) [inline cookbook/security/throttle_failed_login.rst]: I think it'll be easier to follow if you inject the EntityManager, then fetch some fake "AppBundle:LoginAttempt" entity (or something like that). Then we should also show this registered as a service. Finally, we can add a note that we'll leave the LoginAttempt entity logic code to the user.

File: cookbook/security/throttle_failed_login.rst [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+=====================================
+
+Sometimes, it's useful to throttle login attempts when you encounter multiple
+recent failed logins.  This can be useful in combatting brute-force
+password-guessing login attacks.  This can be done by implementing a
+``security.authentication.failure`` listener to log failed login attempts, and
+a security provider decorator to deny login attempts from the same IP address.
+
+Authentication Failure Listener
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (weaverryan) [inline cookbook/security/throttle_failed_login.rst]: we use `-----` for the first non-title headline

File: cookbook/security/throttle_failed_login.rst [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                $event->getToken(),
+                $this->requestStack->getCurrentRequest()
+            );
+        }
+    }
+
+Now you are logging all the relevant details about every failed login attempt.
+
+Security Provider Decorator
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (weaverryan) [inline cookbook/security/throttle_failed_login.rst]: same here!

File: cookbook/security/throttle_failed_login.rst [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+        public function __construct(RequestStack $requestStack, EntityRepositoryInterface $repository)
+        {
+            $this->repository = $repository;
+            $this->requestStack = $requestStack;
+        }
+
+        public function onAuthenticationFailureEvent(AuthenticationFailureEvent $event)
+        {
+            $this->repository->logAuthenicationFailure(
```
Comment (weaverryan) [inline cookbook/security/throttle_failed_login.rst]: typo on logAuthenticationFailure :)

File: cookbook/security/throttle_failed_login.rst [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+        public function __construct(AuthenticationProviderInterface $provider, $throttleService)
+        {
+            $this->decoratedProvider = $provider;
+            $this->throttleService = $throttleService;
+        }
+
+        public function authenticate(TokenInterface $token)
+        {
+            if ($this->throttleService->isThrottled($token)) {
```
Comment (weaverryan) [inline cookbook/security/throttle_failed_login.rst]: I think it'll be more clear if we don't inject some un-shown "throttleService", but instead put in a little "fake" logic to show what they should do. And on that note, what would be a good example? It seems like the "Token" is kind of useless. If I'm throttling, wouldn't it be based on something like the IP address? Perhaps we need to inject the request_stack here to make this more realistic.

**Comment by weaverryan** [comment]
@kevintweber don't worry about the `request_stack` thing. Actually, if you could do everything with the `request` service that would be ideal. Then I can easily merge into 2.3, then merge up to 2.6 and change to `request_stack` there.

File: components/security/authentication.rst (hunk 2 above)
Comment (EAnushan) [inline components/security/authentication.rst]: As far as I can tell, I don't think this is true. I just tried with Symfony 3 and an authenticator implementing `SimplePreAuthenticatorInterface`. `security.authentication.success` is called only the first time you are authenticated. Thus, if you are already logged in, none of the authentication events will fire.

**Comment by weaverryan** [comment]
Ping @kevintweber! Will you have some time to finish this?

Thanks!

**Comment by kevintweber** [comment]
@weaverryan  Yeah, I'll do it now.  I'm going to remove the cookbook entry because it's rather contrived, and no application should implement such functionality.  It's just a bad example.

**Comment by kevintweber** [comment]
The text is ready to go.  I cannot see the merge conflicts, so you may need to massage it a bit.  (And sorry about the huge delay.)  If you want me to squash the commits, just say so.

**Comment by weaverryan** [comment]
You did a VERY nice job explaining this - a really nice addition. Thanks so much Kevin!

**Pull request merged**